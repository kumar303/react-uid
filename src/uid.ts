export const generateUID = () => {
  let counter = 1;

  const map = new WeakMap<any, number>();

  const uid = (item: any, index?: number): string => {
    if (
      typeof item === 'number' ||
      typeof item === 'string'
    ) {
      return index ? `idx-${index}` : `val-${item}`;
    }

    if (!map.has(item)) {
      map.set(item, counter++);
      return uid(item)
    }
    return 'uid' + map.get(item);
  };

  return uid;
}

export const generatedUID = generateUID();