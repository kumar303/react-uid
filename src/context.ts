import * as React from 'react';
import {generateUID} from "./uid";

export type IdSourceType = {
  value: number,
  prefix: string,
  uid: (item: any, index?: number) => string,
}

export const createSource = (prefix = 'ruid-'): IdSourceType => ({
  value: 1,
  prefix: prefix,
  uid: generateUID()
});

export interface UIDProps {
  name?: (n: string | number) => string,
  idSource?: IdSourceType,
  children: (id: string, uid: (item: any, index?: number) => string) => React.ReactNode
}

export const counter = createSource();

export const source = React.createContext(createSource());

export const getId = (source: IdSourceType) => source.value++;
export const getPrefix = (source?: IdSourceType) => source ? source.prefix : 'ruid-';