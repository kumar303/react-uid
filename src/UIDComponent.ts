import * as React from 'react';
import {counter, getId, getPrefix, UIDProps} from "./context";

// --------------------------------------------

export class UID extends React.Component<UIDProps> {
  state = {
    quartz: this.props.idSource || counter,
    prefix: getPrefix(this.props.idSource),
    id: getId(this.props.idSource || counter)
  };

  prefixId(id: number | string) {
    const uid = (this.state.prefix + id);
    return String(this.props.name ? this.props.name(uid) : uid);
  }

  uid = (item: any) => this.prefixId(this.state.id + '-' + this.state.quartz.uid(item));

  render() {
    const {children} = this.props;
    const {id} = this.state;
    return children(this.prefixId(id), this.uid)
  }
}