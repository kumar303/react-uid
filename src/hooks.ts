import * as React from "react";
import {counter, getId, getPrefix, IdSourceType, source} from "./context";

type SeedGenerator = (id: any) => string;

const generateUID = (context: IdSourceType) => {
  const quartz = context || counter;
  const prefix = getPrefix(quartz);
  const id = getId(quartz);
  const uid = prefix + id;
  const gen = (item: any) => uid + quartz.uid(item);
  return {uid, gen};
};

const useUIDState = () => {
  if (process.env.NODE_ENV !== "production") {
    if (!('useContext' in React)) {
      throw new Error('Hooks API requires React 16.8+');
    }
  }

  return React.useState(generateUID(React.useContext(source)))
};

export const useUID = (): string => {
  const [{uid}] = useUIDState();
  return uid;
};

export const useUIDSeed = (): SeedGenerator => {
  const [{gen}] = useUIDState();
  return gen;
};