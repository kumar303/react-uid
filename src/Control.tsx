import * as React from 'react';
import {createSource, UIDProps, source} from "./context";
import {UID} from "./UIDComponent";

interface WithPrefix {
  prefix?: string;
}

export const UIDReset: React.SFC<WithPrefix> = ({children, prefix = ''}) => (
  <source.Provider value={createSource(prefix)}>{children}</source.Provider>
);

export const UIDFork: React.SFC<WithPrefix> = ({children, prefix = ''}) => (
  <UIDConsumer>
    {(id) => (
      <source.Provider value={createSource(id + '-' + prefix)}>
        {children}
      </source.Provider>
    )}
  </UIDConsumer>
);

export const UIDConsumer: React.SFC<UIDProps> = ({name, children}) => (
  <source.Consumer>
    {(value) => (
      <UID name={name} idSource={value} children={children}/>
    )}
  </source.Consumer>
);