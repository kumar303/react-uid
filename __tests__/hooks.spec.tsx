import * as React from 'react';
import * as TestRenderer from 'react-test-renderer';
import {UIDReset, UIDFork, UIDConsumer, useUID, useUIDSeed} from "../src";


describe('uid hooks', () => {
  it('test uid', () => {
    const Target = () => {
      const id = useUID();
      const seed = useUIDSeed();
      return <i>id1: {id} id2:{seed('second-')}</i>;
    }

    const Component = (
      <UIDReset prefix={"(test-uid)"}>
        <Target/>
        <Target/>
        ---fork1:
        <UIDFork prefix={"fork1:"}>
          <Target/>
        </UIDFork>
        ---
        <Target/>
        ---fork2:
        <UIDFork>
          <Target/>
        </UIDFork>
        ---
        <UIDConsumer>
          {(uid) => <i>{uid}</i>}
        </UIDConsumer>
      </UIDReset>
    );

    const testRenderer = TestRenderer.create(Component);
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });
});