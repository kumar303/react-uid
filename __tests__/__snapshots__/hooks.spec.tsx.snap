// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uid hooks test uid 1`] = `
Array [
  <i>
    id1: 
    (test-uid)1
     id2:
    (test-uid)2val-second-
  </i>,
  <i>
    id1: 
    (test-uid)3
     id2:
    (test-uid)4val-second-
  </i>,
  "---fork1:",
  <i>
    id1: 
    (test-uid)5-fork1:1
     id2:
    (test-uid)5-fork1:2val-second-
  </i>,
  "---",
  <i>
    id1: 
    (test-uid)6
     id2:
    (test-uid)7val-second-
  </i>,
  "---fork2:",
  <i>
    id1: 
    (test-uid)8-1
     id2:
    (test-uid)8-2val-second-
  </i>,
  "---",
  <i>
    (test-uid)9
  </i>,
]
`;
